1) Создал ветку sprint_8-solution-time-and-duration

ViktorSalk opened the pull request:
1) Добавил новую функциональность: трекер задач ожет расставлять задачи по приоритету и проверять, не пересекаются ли они по времени выполнения.

2) Добавил новые поля и методы в задачи: duration, LocalDateTime, getEndTime(). 3) Добавил валидацию во время создания или изменения задач и подзадач. 4) Добавил тесты.
5) Добавил базовый класс TaskManager с тестами для InMemoryTaskManager и FileBackedTaskManager 6) Добавил тесты HistoryManagerTest для HistoryManager 7) Использовал тип Optional, чтобы возвращать результат методов поиска задачи/подзадачи по id. 8) Добавил тесты InMemoryTaskManager для новой фичи по Optional.

Viktor3642 committed:
1) Добавил новую функциональность: трекер задач ожет расставлять задачи по приоритету и проверять, не пересекаются ли они по времени выполнения.
2) Добавил новые поля и методы в задачи: duration, LocalDateTime, getEndTime().
3) Добавил валидацию во время создания или изменения задач и подзадач.
4) Добавил тесты.
5) Добавил базовый класс TaskManager с тестами для InMemoryTaskManager и FileBackedTaskManager
6) Добавил тесты HistoryManagerTest для HistoryManager
7) Использовал тип Optional, чтобы возвращать результат методов поиска задачи/подзадачи по id.
8) Добавил тесты InMemoryTaskManager для новой фичи по Optional.

Viktor3642 committed:
1) Убрал неиспользуемый импорт

afomkina commented on src/main/java/manager/FileBackedTaskManager.java:
```diff
@@ -36,7 +38,7 @@ protected void loadFromFile() {
                         addEpic((Epic) task);
                     } else if (task instanceof Subtask) {
```
Нужно избавиться от instanceof, так как он не соответствует парадигме ООП. Для этого нужно использовать enum с типом задач

ViktorSalk replied:
В FileBackedTaskManager избавился от instanceof. Использовал enum с типом задач.

afomkina commented on src/main/java/history/HistoryManagerImpl.java:
```diff
@@ -0,0 +1,38 @@
+
+    @Override
+    public void add(Task task) {
+        if (task != null && !history.stream().anyMatch(t -> t.getId() == task.getId())) {
```
- Лучше использовать `noneMatch` вместо инверсии с `anyMatch`
- Нужно переименовать однобуквенную переменную

ViktorSalk replied:
В HistoryManagerImpl использовал noneMatch вместо инверсии с anyMatch, переименовал однобуквенную переменную в existingTask.

afomkina commented on src/main/java/manager/FileBackedTaskManager.java:
```diff
@@ -149,15 +138,18 @@ protected void save() {
     // Преобразуйте задачу в строковое представление CSV
     private String toString(Task task) {
         if (task instanceof Subtask) {
```
Нужно избавиться от instanceof, так как он не соответствует парадигме ООП. Для этого нужно использовать enum с типом задач

ViktorSalk replied:
В FileBackedTaskManager избавился от instanceof. Использовал enum.

afomkina commented on test/java/history/HistoryManagerTest.java:
```diff
@@ -0,0 +1,101 @@
+
+import static org.junit.jupiter.api.Assertions.*;
+
+public class HistoryManagerTest {
```
Предлагаю немного подекомпозировать тесты и разбить из на логические составляющие. Это можно сделать с помощью [@Nested классов](https://www.petrikainulainen.net/programming/testing/junit-5-tutorial-writing-nested-tests/). Так код станет более читаемым и его будет проще поддерживать

ViktorSalk replied:
Декомпозировал тесты и разбил их на логические составляющие.

afomkina commented on test/java/history/HistoryManagerTest.java:
```diff
@@ -0,0 +1,101 @@
+
+    @Test
+    public void testRemoveBeginning() {
+        Task task1 = new Task("Задача 1", "Описание 1");
```
Нужно максимально избавиться от дублирований в тесте. Создание  повторяющихся объектов задач, предлагаю вынести в отдельный утильный класс и сделать статическим методом (assignTask, assignEpic...), чтобы его можно быть переиспользовать в каждом из тестов

ViktorSalk replied:
Спасибо =)) Убрал дублирования в тесте.

afomkina commented on test/java/manager/InMemoryTaskManagerTest.java:
```diff
@@ -14,13 +14,15 @@
+    }
+    @Override
+    protected InMemoryTaskManager createTaskManager() {
+        return new InMemoryTaskManager();
```
Должен создаваться с помощью фабрики Managers

ViktorSalk replied:
В HistoryManagerTest новые экземпляры создаются с помощью фабрики Managers

afomkina commented on test/java/manager/InMemoryTaskManagerTest.java:
```diff
@@ -14,13 +14,15 @@
     @BeforeEach
-    void setUp() {
-        taskManager = Managers.getDefault();
+    public void setUp() {
```
Этот метод нужно удалить так как присвоение происходит в `TaskManagerTest`

ViktorSalk replied:
В InMemoryTaskManagerTest удалил setUp так как присвоение происходит в TaskManagerTest.

afomkina commented on test/java/manager/InMemoryTaskManagerTest.java:
```diff
@@ -14,13 +14,15 @@
+    @Override
+    protected InMemoryTaskManager createTaskManager() {
+        return new InMemoryTaskManager();
     }
```
Все общие тесты, которые есть в классе ниже должны находиться в `TaskManagerTest`, отсюда их нужно удалить

ViktorSalk replied:
Удалил лишние тесты уже имеющиеся в TaskManagerTest

Viktor3642 committed:
1) В FileBackedTaskManager избавился от instanceof, так как он не соответствует парадигме ООП. Использовал enum с типом задач.
2) В HistoryManagerImpl использовал noneMatch вместо инверсии с anyMatch, переименовал однобуквенную переменную в existingTask.
3) Декомпозировал тесты и разбил их на логические составляющие.
4) Убрал дублирования в тесте.
5) В HistoryManagerTest новые экземпляры создаются с помощью фабрики Managers
6) В InMemoryTaskManagerTest удалил setUp так как присвоение происходит в TaskManagerTest.
7) Удалил лишние тесты уже имеющиеся в TaskManagerTest

afomkina commented on test/java/history/HistoryManagerTest.java:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    @Nested
+    class AddTaskTests { // Тесты добавления задач в историю
```
👍

afomkina commented on test/java/manager/TimeSlotManagerTest.java:
```diff
@@ -0,0 +1,95 @@
+import static org.junit.jupiter.api.Assertions.*;
+
+public class TimeSlotManagerTest {
+    private InMemoryTaskManager taskManager;
```
Должен быть интерфейсом `TaskManager`

ViktorSalk replied:
Изменил тип на TaskManager

afomkina commented on test/java/manager/TimeSlotManagerTest.java:
```diff
@@ -0,0 +1,95 @@
+
+    @BeforeEach
+    public void setUp() {
+        taskManager = new InMemoryTaskManager();
```
Должен создаваться с помощью метода из фабрики `Managers`

ViktorSalk replied:
Заменил, теперь в setUp новые экземпляры создаются с помощью фабрики Managers

afomkina commented on src/main/java/manager/Managers.java:
```diff
@@ -4,6 +4,15 @@
 public class Managers {
+    private static InMemoryTaskManager taskManager;
+
+    public static InMemoryTaskManager getInMemoryTaskManager() {
```
Есть метод `getDefault`, который возвращает `InMemoryTaskManager`, нужно использовать его, а этот метод удалить

ViktorSalk replied:
Удалил getInMemoryTaskManager оставив метод getDefault, возвращающий InMemoryTaskManager.

Viktor3642 committed:
1) В TimeSlotManagerTest изменен тип на TaskManager
2) В TimeSlotManagerTest в setUp новые экземпляры создаются с помощью фабрики Managers
3) В Managers удалил getInMemoryTaskManager оставив метод getDefault, возвращающий InMemoryTaskManager.
